Add Git helper method to check if a path results ignored

mokagio opened the pull request:
Adds a bit of security to the secrets management by preventing a user from decrypting a secret to a location that is not `.gitgnored`.

It covers the `configure_apply`, `_update`, and `_add_files_to_copy` actions.

I decided not to edit the `.gitignore` via the automation to leave the user the flexibility to change path or use a pattern in the `.gitignore`.

To test this implementation, you can checkout [this demo PR](https://github.com/Automattic/simplenote-ios/pull/1263) or do something along the lines of:

1. Get a repo on this version of the release toolkit
2. Update the `.configure` with a file to decrypt to a location that's not ignored
3. Run the `configure_apply` and `configure_update` actions to verify they show an error
4. Run the `configure_add_files_to_copy` and pass a destination path not ignored to verify it shows an error

mokagio committed:
Add Git helper method to check if a path results ignored

mokagio committed:
Add more Ruby-ish '?' in is_ignored method name

mokagio committed:
Prevent copying decrypted files to a location that is not ignored by Git

mokagio committed:
Wrap configure_apply in begin-rescue

This way it will show a better output if a failure occurs.

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/configure/configure_apply_action.rb:
```diff
@@ -54,39 +54,43 @@ def self.prepare_repository
       ### Check with the user whether we should overwrite the file, if it exists
       ###
       def self.apply_file(file_reference, force)
```
Unfortunately, the diff is noisy. The only change here was wrapping the logic in a `begin – rescue` block, to gracefully handle the expected error raised if a file is not ignored without dumping the stacktrace on the user

AliSoftware replied:
"Ignore whitespace" to the rescue! 
![image](https://user-images.githubusercontent.com/216089/116390529-6c09d100-a81e-11eb-88d6-a97b8cb1745e.png)

mokagio committed:
Use less implementation-focused test for Git is_ignored? helper

mokagio committed:
Use system in favor of Action.sh to run git check-ignore

A test was giving me an unexpected result with Action.sh.
Using a pure Ruby implementation seems better anyways.

mokagio committed:
Address new Rubocop violations in git_helper_spec

mokagio commented on spec/file_reference_spec.rb:
```diff
@@ -29,11 +29,30 @@
-      expect(File).to receive(:write).with(subject.destination_file_path, 'source contents')
-      subject.apply
+    context 'when the destination is not ignored in Git' do
+      it 'raises' do
+        allow(Fastlane::Helper::GitHelper).to receive(:is_ignored?)
+          .with(path: subject.destination_file_path)
+          .and_return(false)
+
+        expect(FileUtils).to_not receive(:mkdir_p)
+        expect(subject).to_not receive(:source_contents)
+        expect(File).to_not receive(:write)
+        expect { subject.apply }.to raise_error(StandardError)
+      end
+    end
+
+    context 'when the destination is ignored in Git' do
+      it 'copies the source to the destination' do
+        allow(Fastlane::Helper::GitHelper).to receive(:is_ignored?)
+          .with(path: subject.destination_file_path)
+          .and_return(true)
+
+        allow(FileUtils).to receive(:mkdir_p)
+        allow(subject).to receive(:source_contents).and_return('source contents')
+        expect(File).to receive(:write).with(subject.destination_file_path, 'source contents')
+        subject.apply
+      end
```
I don't like how heavy on runtime mocking these tests are, they're more about the implementation (call `File.write`) than the actual behavior.

On the other hand, this was simpler to achieve than putting in place a temp folders etc.

mokagio committed:
Add check for destination ignored in Git when adding a file, too

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/configure_helper.rb:
```diff
@@ -225,6 +225,10 @@ def self.add_file(params)
 
         UI.user_error!('You must pass a `destination` to `add_file`') unless params[:destination]
```
I was tempted to refactor this method to use keyword arguments instead of an hash with parameters. I decided to save it for a later refactor.

cc @AliSoftware in case that's something that's already on your TODO list.

mokagio commented on spec/git_helper_spec.rb:
```diff
@@ -90,4 +91,58 @@
+        dummy_file_path: path,
+        in_gitignore: false
+      )
+      expect(Fastlane::Helper::GitHelper.is_ignored?(path: path)).to be false
```
I could have used `described_class` here like in the spec above, but I decided not to because the rest of the file doesn't.

(Also, I remembered about `described_class` after writing these tests... 😅 )

AliSoftware replied:
We still haven't configured CircleCI to run rubocop on each PR (something I really need to take care of before we start reintroducing too many new violations via all our recent updates after my previous cleanup!!) but I know that this would have been flagged by rubocop…

mokagio replied:
🤔 I actually did run Rubocop before submitting my PR (and found some issues, see 79516ac)

What were you expecting it to flag? Right now, it's all green:

(_I tried to upload a screenshot but I'm getting "Something went really wrong" from the editor._ _I also tried to screenshot and upload that..._ 🤦 😅 ) 




I may be able to do it tomorrow / Friday. Were you thinking something simple like "run on every file that changed" or more advanced like comment in line with the violations?

Oh 🤔 Now that I think of it, maybe we should setup Hound?

AliSoftware replied:
> What were you expecting it to flag? Right now, it's all green:

I was expecting [this rule](https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/DescribedClass) to be triggered 🤔 

> Oh 🤔 Now that I think of it, maybe we should setup Hound?

Woof!! (That means "definitively!!" in case you don't speak dog)

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/configure_helper.rb:
```diff
@@ -225,6 +225,10 @@ def self.add_file(params)
         UI.user_error!('You must pass a `destination` to `add_file`') unless params[:destination]
 
+        unless Fastlane::Helper::GitHelper.is_ignored?(path: params[:destination])
+          UI.user_error! "Attempted to add a file to a location which is not ignored under Git (#{params[:destination]})"
```
```suggestion
          UI.user_error! "Attempted to add a file to a location which is not ignored under Git (#{params[:destination]}). Please edit your `.gitignore` manually to fix this."
```

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/git_helper.rb:
```diff
@@ -159,6 +159,17 @@ def self.ensure_on_branch!(branch_name)
+      #
+      # @return [Bool] True if the given path is ignored, false otherwise.
+      def self.is_ignored?(path:)
+        system("git check-ignore #{path}")
+        $CHILD_STATUS.exitstatus == 0
```
```suggestion
        sh('git', 'check-ignore', path) { |status, _, _| status.success? }
```

(Not tested, so worth checking my syntax, but according to https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/helper/sh_helper.rb#L11-L28 I think that's the right one)

mokagio replied:
Thanks. That looks neat. Is there any benefit to using that syntax compared to this one?

AliSoftware replied:
`sh` is a method provided by Fastlane, so it's nicer to use it because it wraps the `system` call around a couple of things, including reporting results in logs via `UI.xxx` messages, capturing errors for reporting too, disabling real execution by default when running it in unit tests…

mokagio replied:
I implemented this in b9d543a 👍 

I'm not convinced by your argument, though 😄 

For the purpose of this method, I think using vanilla Ruby was better. It doesn't depend on anything external so it's less likely to break because there's one less moving part. Also, I had to [update the tests](https://github.com/wordpress-mobile/release-toolkit/pull/242/commits/b9d543a0a5a3d60380c36c0cab4750de7517c920#diff-7aed954270eabe63c0117e184d02a920e7c50e090136cd8a33fe121bd7e87ca9R95-R97) to account for the usage of `Action.sh` by calling `allow(FastlaneCore::Helper).to receive(:sh_enabled?).and_return(true)`, which adds an avoidable dependency on the code implementation to tests that had no dependency to it before that.

Neither of my points is very strong, though. It's possible that using Fastlane will actually prevent future bugs coming from Ruby itself, it also delegates the non-trivial job of properly shelling out to a codebase used by millions of people, which should make it quite stable at that.

As for the tests, well the dependency is minor and it's nice that is not about `sh` itself but rather something internal. That is, if we were to use a different helper method or drop `Action.sh` altogether, the tests will likely still pass.

**I'm happy to go with this change, because, after all, this code is deeply embedded in the  Fastlane ecosystem, so it makes sense to leverage some of its utilities.** 😄 🚗 💨

---

@AliSoftware, you mention "reporting results in logs via UI.xxx messages, capturing errors for reporting too".

How does that work? I tried to run the action [in my test PR with these changes](https://github.com/Automattic/simplenote-ios/pull/1263) with `--verbose` but I didn't get anything extra in the logs from this `sh` call 🤔

AliSoftware replied:
Yeah I see your point it's debatable whether to use `sh` or not.

I always try to avoid calling `Kernel#system` directly tho as it might be open to injection iirc (due to it running a subshell and going thru shell expansion), so that's why `popen2e` — which is what `sh` uses — and its family of similar `popen*` functions (which directly spawns the subprocess without an intermediate subshell) is recommended instead, and has the advantages of:

1. not be vulnerable to shell expansion like `system` is, and instead takes an array or arguments and spawn the subprocess and pass it its argv list directly (without going thru a shell re-splitting the args and having to handle quoting and escapes and all);
2. be able to capture both `stdout` and `stderr` (and exit `status`) all at once or separately, unlike `system`;
3. Not rely on a global variable — like `$CHILD_STATUS` — to read the result.

If you're curious about how sh is implemented and what call it wraps to end up invoking the external scripts or command lines, you can take a look [here](https://github.com/fastlane/fastlane/blob/e874a47c6e2e0e61590a03d3b71e75e5a505d1ce/fastlane/lib/fastlane/helper/sh_helper.rb).

In the end for this use case it might not end up being a big deal, but better be consistent and avoid spawning an intermediate shell (that evaluates the command and args and quotes and escapes to split that into args, like with `system`) for nothing and invoke the command directly instead (like with `popen2e` or `sh`) 😇

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/models/file_reference.rb:
```diff
@@ -42,7 +42,18 @@ def update
       def apply
+        # Only decrypt the file if the destination is ignored in Git
+        unless Fastlane::Helper::GitHelper.is_ignored?(path: destination_file_path)
+          raise StandardError.new "Attempted to decrypt #{file} to #{destination_file_path} which is not ignored under Git"
```
Surprised that rubocop doesn't suggest you ere to directly use the String literal without `StandardError.new`, as this is the more idiomatic way iinm. That file might be excluded from rubocop though (because I never had time to finish fixing all the existing violations back then)
```suggestion
          raise "Attempted to decrypt #{file} to #{destination_file_path} which is not ignored under Git"
```

mokagio replied:
Rubocop _did_ suggest it, once I enabled it in my Vim setup... 😅 

It also pointed out a few more violations which I'll address later in this PR, as I'm running out of time today. I'll re-request your review once done.

Thank you for pointing this out 👍

AliSoftware commented on spec/git_helper_spec.rb:
```diff
@@ -1,6 +1,7 @@
 require 'tmpdir'
 require_relative './spec_helper'
 
+# rubocop:disable Metrics/BlockLength
```
I'm surprised this is expected, as I am pretty sure I explicitly excluded `*_spec.rb` files from this rule in `.rubocop.yml` already 🤔

mokagio replied:
You are right. Turns out my `develop` wasn't up to date when created this branch. I rebased and it didn't warn me. 

Fixed up in 830d39c.

AliSoftware commented on spec/git_helper_spec.rb:
```diff
@@ -90,4 +91,58 @@
+    end
+  end
+end
+# rubocop:enable Metrics/BlockLength
```
(don't forget to remove this one if you do remove the disable after my previous comment)

AliSoftware commented on spec/git_helper_spec.rb:
```diff
@@ -90,4 +91,58 @@
+end
+# rubocop:enable Metrics/BlockLength
+
+def setup_git_repo(dummy_file_path:, in_gitignore:, gitignore_committed: false)
```
param naming suggestion:
```suggestion
def setup_git_repo(dummy_file_path:, add_file_to_gitignore:, commit_gitignore: false)
```
(+ adapt implementation accordingly)

AliSoftware reviewed (commented):
Mostly LGTM, just left some nits and improvement suggestions though.

mokagio committed:
Tweak helper method parameters in git_helper_spec

See https://github.com/wordpress-mobile/release-toolkit/pull/242#discussion_r623905782

Thanks @AliSoftware

mokagio committed:
Use simpler raise instead of raise StandardError.new

See
https://github.com/wordpress-mobile/release-toolkit/pull/242#discussion_r623902301

Thanks @AliSoftware.

mokagio committed:
Improve failure message when file decrypt destination is not ignored

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

mokagio committed:
Replace to_not with not_to in file_reference_spec.rb as per RuboCop

mokagio committed:
Extract some setup code in a before block in file_reference_spec

This was to comply with the RuboCop rule on the example length, which
wants examples at 5 lines or less.

To be honest, I feel like I'm cheating a bit here and I feel maybe the
test is less readable as a result.

Opened #248 to track this discussion.

mokagio committed:
Revert "Extract some setup code in a before block in file_reference_spec"

This reverts commit b63a2a30ee5de849c3146f29fd9692b0d28913c5.

mokagio committed:
Use Action.sh instead of system to lean into the Fastlane integration

As suggested here:
https://github.com/wordpress-mobile/release-toolkit/pull/242#discussion_r623900592

Thanks @AliSoftware.

mokagio commented:
@AliSoftware I think I addressed all your great comments. I'd appreciate a second look 🙏 👀

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/git_helper.rb:
```diff
@@ -165,6 +165,16 @@ def self.ensure_on_branch!(branch_name)
+      # Checks whether a given path is ignored by Git, relying on Git's
+      # `check-ignore` under the hood.
+      #
+      # @param [String] path The path to check against the the `.gitignore`
```
```suggestion
      # @param [String] path The path to check against `.gitignore`
```

AliSoftware commented on spec/git_helper_spec.rb:
```diff
@@ -90,4 +90,61 @@
+  end
+end
+
+def setup_git_repo(dummy_file_path:, add_file_to_gitignore:, commit_gitignore: false)
+  `git init`
+  `touch .gitignore`
+  `git add .gitignore && git commit -m 'Add .gitignore'`
+
+  `echo abc > #{dummy_file_path}`
+
+  # no point in commiting the gitignore if the file shouldn't be in it
+  return unless add_file_to_gitignore
+
+  `echo #{dummy_file_path} > .gitignore`
+  `git add .gitignore && git commit -m 'Update .gitignore'` if commit_gitignore
```
First time I read this I thought "it would be nice and safer if all this was done in a temporary directory". Then I went back to read the whole spec file and saw that the tmpdir switch is done via `before` and `after` on the whole spec, so should be ok after all, just thought I'd leave that comment around so that other reviewers aren't tricked by same train of thought 😛

mokagio replied:
When I wrote the first version of this test, it took me at least 5 minutes or beard scratching to figure out why it wasn't working, which was because of the temp dir being set in the `before`.

There's some lessons here for me, nothing new, but stuff that bears repeating:

- Always read (or at least skim) a test file before starting to work on it
- Beware of global setup code, it's sneaky and easy to overlook

AliSoftware replied:
> There's some lessons here for me, nothing new, but stuff that bears repeating:

you forgot:

3. Keep your beard, it's always handy to have it around to be able to scratch it.

AliSoftware reviewed (commented):
Code LGTM 👍 
Haven't taken the time to run the test steps yet though, so will circle back later today to test it before approving

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/models/file_reference.rb:
```diff
@@ -42,7 +42,18 @@ def update
       def apply
+        # Only decrypt the file if the destination is ignored in Git
+        unless Fastlane::Helper::GitHelper.is_ignored?(path: destination_file_path)
+          raise "Attempted to decrypt #{file} to #{destination_file_path} which is not ignored under Git"
```
```suggestion
          raise "Attempted to decrypt #{file} to #{destination_file_path} which is not ignored under Git. Please either edit your `.configure` file to use an already-ignored destination, or add that destination `.gitignore` manually to fix this."
```

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/configure_helper.rb:
```diff
@@ -225,6 +225,10 @@ def self.add_file(params)
         UI.user_error!('You must pass a `destination` to `add_file`') unless params[:destination]
 
+        unless Fastlane::Helper::GitHelper.is_ignored?(path: params[:destination])
+          UI.user_error! "Attempted to add a file to a location which is not ignored under Git (#{params[:destination]}). Please edit your `.gitignore` manually to fix this."
```
```suggestion
          UI.user_error! "Attempted to add a file to a location which is not ignored under Git (#{params[:destination]}). Please either edit your `.configure` file to use an already-ignored destination, or add that destination `.gitignore` manually to fix this."
```
I know it's a bit lengthy, but I prefer more hints/documentation than having people (including future us) losing time wondering what's the expected fix 🙂 

PS: Feel free to reword, or introduce a `\n` somewhere in the message, and/or use another ruby string syntax (e.g. heredoc) to format this in a way that makes rubocop happy if needed.

AliSoftware reviewed (approved):
Just added 2 wording nit/suggestions for better error messages; otherwise LGTM and works as expected 👍 

I even tried to trick the script by using `.configure-files/../foo.json` as the destination to see if it would fail see the `../` trick, but it saw right through me and didn't let me have it 👌 

<details><summary>Couldn't outsmart it! 😸 </summary>

```
$ fl run configure_apply
[15:47:29]: -----------------------------
[15:47:29]: --- Step: configure_apply ---
[15:47:29]: -----------------------------
[15:47:29]: $ git check-ignore /Users/olivier/Documents/Dev/apps/woocommerce-ios/.configure-files/../app_store_connect_fastlane_api_key.json
[15:47:29]: Exit status of command 'git check-ignore /Users/olivier/Documents/Dev/apps/woocommerce-ios/.configure-files/../app_store_connect_fastlane_api_key.json' was 1 instead of 0.

[!] Attempted to decrypt iOS/app_store_connect_fastlane_api_key.json to /Users/olivier/Documents/Dev/apps/woocommerce-ios/.configure-files/../app_store_connect_fastlane_api_key.json which is not ignored under Git
```
```
$ fl run configure_add_files_to_copy
[15:48:09]: -----------------------------------------
[15:48:09]: --- Step: configure_add_files_to_copy ---
[15:48:09]: -----------------------------------------
[15:48:09]: Do you want to specify additional files that should be copied from the secrets repository into your project? (y/n)
y
[15:48:16]: -----------------------------------------------------------------------------------------
[15:48:16]: --- Please provide the location of the source file relative to the secrets repository ---
[15:48:16]: -----------------------------------------------------------------------------------------
[15:48:16]: Example: google-services.json
[15:48:16]: Source File Path:
iOS/tracks-demo/Secrets.swift
[15:48:59]: -------------------------------------------------------------------------------
[15:48:59]: --- Please provide the destination of the file relative to the project root ---
[15:48:59]: -------------------------------------------------------------------------------
[15:48:59]: Example: WordPress/google-services.json
[15:48:59]: Destination File Path:
.configure-files/../whoops.swift
[15:49:15]: Encrypt file?: (y/n)
y
[15:49:16]: $ git check-ignore .configure-files/../whoops.swift
[15:49:17]: Exit status of command 'git check-ignore .configure-files/../whoops.swift' was 1 instead of 0.


[!] Attempted to add a file to a location which is not ignored under Git (.configure-files/../whoops.swift). Please edit your `.gitignore` manually to fix this.
```
</details>

mokagio committed:
Fix typo in git_helper.rb documentation

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

mokagio committed:
Improve resolution instructions in the  "decrypted file not ignored" error message

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>
